### ISSUE TITLE
Polymorphism and deserialization of lists (JSON)

### ISSUE THREAD

**Issue by palkatap** [Kotlin/kotlinx.serialization#2278]
**To Reproduce**
I created the following classes:

```kotlin
@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("filterType")
sealed class Base

@Serializable
@SerialName("FIRST")
data class FirstClass(
  val b: Double = 34.0
): Base()
```
this deserialization code throws exception:
```kotlin
Json.decodeFromString<Base>("""{"filterType":"FIRST","b":14.0}""")
```

> Polymorphic serializer was not found for missing class discriminator ('null')

Actually in my case the **b** value comes as string (like {"filterType":"FIRST","b":"14.0"}) and I want to convert it to double.
For this I created a KSerializer:
```kotlin
object DoubleSerializer: KSerializer<Double> {
  override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("DoubleB", PrimitiveKind.STRING)

  override fun deserialize(decoder: Decoder): Double =
    decoder.decodeString().toDouble()

  override fun serialize(encoder: Encoder, value: Double) =
    encoder.encodeString(value.toBigDecimal().toPlainString())
}
```
I also adjusted the FirstClass:
```kotlin
@Serializable
@SerialName("FIRST")
data class FirstClass(
  @Serializable(with = DoubleSerializer::class)
  val b: Double = 34.0
): Base()
```
Now the code:
```kotlin
Json.decodeFromString<Base>("""{"filterType":"FIRST","b":"14.0"}""")
```
runs without exception.
So far so good, this is what I need.

My problem actually is that I cannot deserialize a **list** of such objects. The code:
```kotlin
Json.decodeFromString<List<Base>>("""[{"filterType":"FIRST","b":"14.0"}]""")
```
throws JsonDecodingException:
kotlinx.serialization.json.internal.JsonDecodingException: Polymorphic serializer was not found for missing class discriminator ('null')

**Environment**
 - Kotlin version: [e.g. 1.8.10]
 - Library version: [e.g. 1.5.0]
 - Kotlin platforms: [JVM]
 - Gradle version: [7.6]
 - IDE version [IntellijIDEA 2023.1.1]
 - Other relevant context [Windows 10, JDK 11]

**Comment by sandwwraith** [comment]
I think this is the problem with intrinsics that do not record annotations. Does `Json.decodeFromString(serializer(typeOf<List<Base>>()),  """[{"filterType":"FIRST","b":"14.0"}]"""))` working correctly?

**Comment by sandwwraith** [comment]
Also I think there was a similar bug fixed in 1.8.20. Can you check the most recent Kotlin version?

**Comment by palkatap** [comment]
> Also I think there was a similar bug fixed in 1.8.20. Can you check the most recent Kotlin version?

I upgraded to 1.8.20, and now it works. I tried both:
```kotlin
    Json.decodeFromString(serializer(typeOf<List<Base>>()),  """[{"filterType":"FIRST","b":"14.0"}]""")
    Json.decodeFromString<List<Base>>("""[{"filterType":"FIRST","b":"14.0"}]""")
```
Thank you for your help!
